diff --git a/Makefile b/Makefile
index 1ec8ac3fd..57a674feb 100644
--- a/Makefile
+++ b/Makefile
@@ -288,15 +288,15 @@ MAKEFLAGS += -rR
 
 # Make variables (CC, etc...)
 
-AS		= $(CROSS_COMPILE)as
-CC		= $(CROSS_COMPILE)gcc
+AS		= llvm-as
+CC		= llvm-gcc
 LD		= $(CC) -nostdlib
 CPP		= $(CC) -E
-AR		= $(CROSS_COMPILE)ar
-NM		= $(CROSS_COMPILE)nm
-STRIP		= $(CROSS_COMPILE)strip
-OBJCOPY		= $(CROSS_COMPILE)objcopy
-OBJDUMP		= $(CROSS_COMPILE)objdump
+AR		= llvm-ar
+NM		= llvm-nm
+STRIP		= llvm-strip
+OBJCOPY		= llvm-objcopy
+OBJDUMP		= llvm-objdump
 PKG_CONFIG	?= $(CROSS_COMPILE)pkg-config
 AWK		= awk
 GENKSYMS	= scripts/genksyms/genksyms
diff --git a/Makefile.flags b/Makefile.flags
index 97cb4dca2..6aecde8cb 100644
--- a/Makefile.flags
+++ b/Makefile.flags
@@ -26,7 +26,7 @@ CFLAGS += $(call cc-option,-Wstrict-prototypes,)
 CFLAGS += $(call cc-option,-Wunused -Wunused-parameter,)
 CFLAGS += $(call cc-option,-Wunused-function -Wunused-value,)
 CFLAGS += $(call cc-option,-Wmissing-prototypes -Wmissing-declarations,)
-CFLAGS += $(call cc-option,-Wno-format-security,)
+CFLAGS += $(call cc-option,-Wno-format-security -Wno-ignored-optimization-argument -Wno-implicit-function-declaration,)
 # warn about C99 declaration after statement
 CFLAGS += $(call cc-option,-Wdeclaration-after-statement,)
 # If you want to add more -Wsomething above, make sure that it is
diff --git a/include/libbb.h b/include/libbb.h
index 01cdb1bdc..41189ef5d 100644
--- a/include/libbb.h
+++ b/include/libbb.h
@@ -138,13 +138,15 @@
 #ifndef _PATH_GSHADOW
 #define _PATH_GSHADOW "/etc/gshadow"
 #endif
-#if defined __FreeBSD__ || defined __OpenBSD__
+#if defined __FreeBSD__ || defined __OpenBSD__ || defined __APPLE__
 # include <netinet/in.h>
 # include <arpa/inet.h>
 #elif defined __APPLE__
 # include <netinet/in.h>
 #else
+#ifdef HAVE_NET
 # include <arpa/inet.h>
+#endif
 //This breaks on bionic:
 //# if !defined(__socklen_t_defined) && !defined(_SOCKLEN_T_DECLARED)
 ///* We #define socklen_t *after* includes, otherwise we get
@@ -1382,10 +1384,10 @@ uint32_t getopt32long(char **argv, const char *optstring, const char *longopts,
  * By ~2008, OpenBSD 3.4 was changed to survive glibc-like optind = 0
  * (to interpret it as if optreset was set).
  */
-#if 1 /*def __GLIBC__*/
+#ifndef __APPLE__ /*def __GLIBC__*/
 #define GETOPT_RESET() (optind = 0)
 #else /* BSD style */
-#define GETOPT_RESET() (optind = 1)
+#define GETOPT_RESET() (optind = optreset = 1)
 #endif
 
 
@@ -2295,7 +2297,13 @@ extern const char bb_path_wtmp_file[] ALIGN1;
 #define bb_path_motd_file "/etc/motd"
 
 #define bb_dev_null "/dev/null"
+#ifdef __APPLE__
+const char *
+macos_get_self_exe_path (void);
+#define bb_busybox_exec_path macos_get_self_exe_path()
+#else
 extern const char bb_busybox_exec_path[] ALIGN1;
+#endif
 /* allow default system PATH to be extended via CFLAGS */
 #ifndef BB_ADDITIONAL_PATH
 #define BB_ADDITIONAL_PATH ""
diff --git a/include/platform.h b/include/platform.h
index ea0512f36..428326d22 100644
--- a/include/platform.h
+++ b/include/platform.h
@@ -157,8 +157,13 @@
 #include <limits.h>
 #if defined(__digital__) && defined(__unix__)
 # include <sex.h>
-#elif defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__NetBSD__) \
-   || defined(__APPLE__)
+#elif defined(__APPLE__)
+# include <machine/endian.h>
+# include <libkern/OSByteOrder.h>
+# define bswap_16 OSSwapInt16
+# define bswap_32 OSSwapInt32
+# define bswap_64 OSSwapInt64
+#elif defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__NetBSD__)
 # include <sys/resource.h>  /* rlimit */
 # include <machine/endian.h>
 # define bswap_64 __bswap64
@@ -425,6 +430,7 @@ typedef unsigned smalluint;
 #define HAVE_GETLINE 1
 #define HAVE_XTABS 1
 #define HAVE_MNTENT_H 1
+#define HAVE_NET 1
 #define HAVE_NET_ETHERNET_H 1
 #define HAVE_SYS_STATFS_H 1
 #define HAVE_PRINTF_PERCENTM 1
@@ -496,6 +502,8 @@ typedef unsigned smalluint;
 
 #if defined(__APPLE__)
 # undef HAVE_STRCHRNUL
+# undef HAVE_MEMPCPY
+# undef HAVE_MEMRCHR
 #endif
 
 #if defined(__FreeBSD__)
diff --git a/libbb/Kbuild.src b/libbb/Kbuild.src
index cb8d2c2ec..e5b56a54e 100644
--- a/libbb/Kbuild.src
+++ b/libbb/Kbuild.src
@@ -11,6 +11,7 @@ lib-y:=
 INSERT
 
 lib-y += alloc_affinity.o
+lib-y += macos.o
 lib-y += appletlib.o
 lib-y += ask_confirmation.o
 lib-y += bb_askpass.o
diff --git a/libbb/messages.c b/libbb/messages.c
index 6914d5701..ce914d774 100644
--- a/libbb/messages.c
+++ b/libbb/messages.c
@@ -26,8 +26,9 @@ const char bb_msg_standard_input[] ALIGN1 = "standard input";
 const char bb_msg_standard_output[] ALIGN1 = "standard output";
 
 const char bb_hexdigits_upcase[] ALIGN1 = "0123456789ABCDEF";
-
+#ifndef __APPLE__
 const char bb_busybox_exec_path[] ALIGN1 = CONFIG_BUSYBOX_EXEC_PATH;
+#endif
 const char bb_default_login_shell[] ALIGN1 = LIBBB_DEFAULT_LOGIN_SHELL;
 /* util-linux manpage says /sbin:/bin:/usr/sbin:/usr/bin,
  * but I want to save a few bytes here. Check libbb.h before changing! */
diff --git a/scripts/trylink b/scripts/trylink
index 2255deee7..e7dd9f372 100755
--- a/scripts/trylink
+++ b/scripts/trylink
@@ -1,340 +1,19 @@
-#!/bin/sh
+#!/usr/bin/env python3
 
-#debug=true
-debug=false
+import sys
+import os
 
-# Linker flags used:
-#
-# Informational:
-# --warn-common
-# -Map $EXE.map
-# --verbose
-#
-# Optimizations:
-# --sort-common                 reduces padding
-# --sort-section alignment      reduces padding
-# --gc-sections                 throws out unused sections,
-#                               does not work for shared libs
-# -On                           Not used, maybe useful?
-#
-# List of files to link:
-# $l_list                       == --start-group -llib1 -llib2 --end-group
-# --start-group $O_FILES $A_FILES --end-group
-#
-# Shared library link:
-# -shared                       self-explanatory
-# -fPIC                         position-independent code
-# --enable-new-dtags            ?
-# -z,combreloc                  ?
-# -soname="libbusybox.so.$BB_VER"
-# --undefined=lbb_main          Seed name to start pulling from
-#                               (otherwise we'll need --whole-archive)
-# -static                       Not used, but may be useful! manpage:
-#                               "... This option can be used with -shared.
-#                               Doing so means that a shared library
-#                               is being created but that all of the library's
-#                               external references must be resolved by pulling
-#                               in entries from static libraries."
+out, cc, cflags, ldflags, obj, lib = sys.argv[1:7]
 
+def nonempty(filename):
+    try:
+        return os.stat(filename).st_size > 0
+    except Exception as e:
+        print(filename, e)
+        return False
 
-try() {
-    printf "%s\n" "Output of:" >$EXE.out
-    printf "%s\n" "$*" >>$EXE.out
-    printf "%s\n" "==========" >>$EXE.out
-    $debug && echo "Trying: $*"
-    $@ >>$EXE.out 2>&1
-    return $?
-}
 
-check_cc() {
-    tempname="$(mktemp tmp.XXXXXXXXXX)"
-    echo "int main(int argc,char**argv){return argv?argc:0;}" >"$tempname".c
-    # Can use "-o /dev/null", but older gcc tend to *unlink it* on failure! :(
-    # Was using "-xc /dev/null", but we need a valid C program.
-    $CC $CFLAGS $LDFLAGS $1 "$tempname".c -o "$tempname" >/dev/null 2>&1
-    exitcode=$?
-    rm -f "$tempname" "$tempname".c "$tempname".o
-    return $exitcode
-}
+obj = [fn for fn in obj.split() if nonempty(fn)]
+lib = [fn for fn in lib.split() if nonempty(fn)]
 
-check_libc_is_glibc() {
-    tempname="$(mktemp tmp.XXXXXXXXXX)"
-    echo "\
-	#include <stdlib.h>
-	/* Apparently uclibc defines __GLIBC__ (compat trick?). Oh well. */
-	#if defined(__GLIBC__) && !defined(__UCLIBC__)
-	syntax error here
-	#endif
-	" >"$tempname".c
-    ! $CC $CFLAGS "$tempname".c -c -o "$tempname".o >/dev/null 2>&1
-    exitcode=$?
-    rm -f "$tempname" "$tempname".c "$tempname".o
-    return $exitcode
-}
-
-EXE="$1"
-CC="$2"
-CFLAGS="$3"
-LDFLAGS="$4"
-O_FILES="$5"
-A_FILES="$6"
-# We try to drop libraries from LDLIBS if build works without them,
-# but ones from CONFIG_EXTRA_LDLIBS are always linked in.
-# (For example, musl has stub utmp implementation, and if you link with
-# a real utmp library in LDLIBS, dropping it "works" but resulting binary
-# does not work properly).
-LDLIBS="$7"
-CONFIG_EXTRA_LDLIBS="$8"
-
-# The --sort-section option is not supported by older versions of ld
-SORT_SECTION="-Wl,--sort-section,alignment"
-if ! check_cc "-Wl,--sort-section,alignment"; then
-    echo "Your linker does not support --sort-section,alignment"
-    SORT_SECTION=""
-fi
-
-START_GROUP="-Wl,--start-group"
-END_GROUP="-Wl,--end-group"
-INFO_OPTS() {
-	echo "-Wl,--warn-common -Wl,-Map,$EXE.map -Wl,--verbose"
-}
-
-# gold may not support --sort-common (yet)
-SORT_COMMON="-Wl,--sort-common"
-if ! check_cc "-Wl,--sort-common"; then
-    echo "Your linker does not support --sort-common"
-    SORT_COMMON=""
-fi
-
-# Static linking against glibc produces buggy executables
-# (glibc does not cope well with ld --gc-sections).
-# See sources.redhat.com/bugzilla/show_bug.cgi?id=3400
-# Note that glibc is unsuitable for static linking anyway.
-# We are removing -Wl,--gc-sections from link command line.
-GC_SECTIONS="-Wl,--gc-sections"
-if (. ./.config && test x"$CONFIG_STATIC" = x"y") then
-    if check_libc_is_glibc; then
-	echo "Static linking against glibc, can't use --gc-sections"
-	GC_SECTIONS=""
-    fi
-fi
-# The --gc-sections option is not supported by older versions of ld
-if test -n "$GC_SECTIONS"; then
-    if ! check_cc "$GC_SECTIONS"; then
-	echo "Your linker does not support $GC_SECTIONS"
-	GC_SECTIONS=""
-    fi
-fi
-
-# Sanitize lib list (dups, extra spaces etc)
-LDLIBS=`echo "$LDLIBS" | xargs -n1 | sort | uniq | xargs`
-
-# First link with all libs. If it fails, bail out
-echo "Trying libraries: $LDLIBS"
-# "lib1 lib2 lib3" -> "-llib1 -llib2 -llib3"
-l_list=`echo " $LDLIBS $CONFIG_EXTRA_LDLIBS " | sed -e 's: \([^- ][^ ]*\): -l\1:g' -e 's/^ *//'`
-test x"$l_list" != x"" && l_list="$START_GROUP $l_list $END_GROUP"
-try $CC $CFLAGS $LDFLAGS \
-	-o $EXE \
-	$SORT_COMMON \
-	$SORT_SECTION \
-	$GC_SECTIONS \
-	$START_GROUP $O_FILES $A_FILES $END_GROUP \
-	$l_list \
-|| {
-    echo "Failed: $l_list"
-    cat $EXE.out
-    echo 'Note: if build needs additional libraries, put them in CONFIG_EXTRA_LDLIBS.'
-    echo 'Example: CONFIG_EXTRA_LDLIBS="pthread dl tirpc audit pam"'
-    exit 1
-}
-
-# Now try to remove each lib and build without it.
-# Stop when no lib can be removed.
-while test "$LDLIBS"; do
-    $debug && echo "Trying libraries: $LDLIBS"
-    dropped_non_first_lib=false
-    first_lib=true
-    for one in $LDLIBS; do
-	without_one=`echo " $LDLIBS " | sed "s/ $one / /g" | xargs`
-	# "lib1 lib2 lib3" -> "-llib1 -llib2 -llib3"
-	l_list=`echo " $without_one $CONFIG_EXTRA_LDLIBS " | sed -e 's: \([^- ][^ ]*\): -l\1:g' -e 's/^ *//'`
-	test x"$l_list" != x"" && l_list="$START_GROUP $l_list $END_GROUP"
-	$debug && echo "Trying -l options: '$l_list'"
-	try $CC $CFLAGS $LDFLAGS \
-		-o $EXE \
-		$SORT_COMMON \
-		$SORT_SECTION \
-		$GC_SECTIONS \
-		$START_GROUP $O_FILES $A_FILES $END_GROUP \
-		$l_list
-	if test $? = 0; then
-	    echo " Library $one is not needed, excluding it"
-	    LDLIBS="$without_one"
-	    $first_lib || dropped_non_first_lib=true
-	else
-	    echo " Library $one is needed, can't exclude it (yet)"
-	    first_lib=false
-	fi
-    done
-    # We can stop trying to drop libs if either all libs were needed,
-    # or we excluded only the _first_ few.
-    # (else: we dropped some intermediate lib(s), maybe now we can succeed
-    # in dropping some of the preceding ones)
-    $dropped_non_first_lib || break
-done
-
-# Make the binary with final, minimal list of libs
-echo "Final link with: ${LDLIBS:-<none>}"
-l_list=`echo " $LDLIBS $CONFIG_EXTRA_LDLIBS " | sed -e 's: \([^- ][^ ]*\): -l\1:g' -e 's/^ *//'`
-test x"$l_list" != x"" && l_list="$START_GROUP $l_list $END_GROUP"
-# --verbose gives us gobs of info to stdout (e.g. linker script used)
-if ! test -f busybox_ldscript; then
-    try $CC $CFLAGS $LDFLAGS \
-	    -o $EXE \
-	    $SORT_COMMON \
-	    $SORT_SECTION \
-	    $GC_SECTIONS \
-	    $START_GROUP $O_FILES $A_FILES $END_GROUP \
-	    $l_list \
-	    `INFO_OPTS` \
-    || {
-	cat $EXE.out
-	exit 1
-    }
-else
-    echo "Custom linker script 'busybox_ldscript' found, using it"
-    # Add SORT_BY_ALIGNMENT to linker script (found in $EXE.out):
-    #  .rodata         : { *(.rodata SORT_BY_ALIGNMENT(.rodata.*) .gnu.linkonce.r.*) }
-    #  *(.data SORT_BY_ALIGNMENT(.data.*) .gnu.linkonce.d.*)
-    #  *(.bss SORT_BY_ALIGNMENT(.bss.*) .gnu.linkonce.b.*)
-    # This will eliminate most of the padding (~3kb).
-    # Hmm, "ld --sort-section alignment" should do it too.
-    #
-    # There is a ld hack which is meant to decrease disk usage
-    # at the cost of more RAM usage (??!!) in standard ld script:
-    #  /* Adjust the address for the data segment.  We want to adjust up to
-    #     the same address within the page on the next page up.  */
-    #  . = ALIGN (0x1000) - ((0x1000 - .) & (0x1000 - 1)); . = DATA_SEGMENT_ALIGN (0x1000, 0x1000);
-    # Replace it with:
-    #  . = ALIGN (0x1000); . = DATA_SEGMENT_ALIGN (0x1000, 0x1000);
-    # to unconditionally align .data to the next page boundary,
-    # instead of "next page, plus current offset in this page"
-    try $CC $CFLAGS $LDFLAGS \
-	    -o $EXE \
-	    $SORT_COMMON \
-	    $SORT_SECTION \
-	    $GC_SECTIONS \
-	    -Wl,-T,busybox_ldscript \
-	    $START_GROUP $O_FILES $A_FILES $END_GROUP \
-	    $l_list \
-	    `INFO_OPTS` \
-    || {
-	cat $EXE.out
-	exit 1
-    }
-fi
-
-. ./.config
-
-sharedlib_dir="0_lib"
-
-if test "$CONFIG_BUILD_LIBBUSYBOX" = y; then
-    mkdir "$sharedlib_dir" 2>/dev/null
-    test -d "$sharedlib_dir" || {
-	echo "Cannot make directory $sharedlib_dir"
-	exit 1
-    }
-    ln -s "libbusybox.so.$BB_VER" "$sharedlib_dir"/libbusybox.so 2>/dev/null
-
-    # Yes, "ld -shared -static" is a thing. It's a shared library which is itself static.
-    LBB_STATIC=""
-    test "$CONFIG_FEATURE_LIBBUSYBOX_STATIC" = y && LBB_STATIC="-Wl,-static"
-
-    EXE="$sharedlib_dir/libbusybox.so.${BB_VER}_unstripped"
-    try $CC $CFLAGS $LDFLAGS \
-	    -o $EXE \
-	    -shared -fPIC $LBB_STATIC \
-	    -Wl,--enable-new-dtags \
-	    -Wl,-z,combreloc \
-	    -Wl,-soname="libbusybox.so.$BB_VER" \
-	    -Wl,--undefined=lbb_main \
-	    $SORT_COMMON \
-	    $SORT_SECTION \
-	    $START_GROUP $A_FILES $END_GROUP \
-	    $l_list \
-	    `INFO_OPTS` \
-    || {
-	echo "Linking $EXE failed"
-	cat $EXE.out
-	exit 1
-    }
-    $STRIP -s --remove-section=.note --remove-section=.comment $EXE -o "$sharedlib_dir/libbusybox.so.$BB_VER"
-    chmod a+x "$sharedlib_dir/libbusybox.so.$BB_VER"
-    echo "libbusybox: $sharedlib_dir/libbusybox.so.$BB_VER"
-fi
-
-if test "$CONFIG_FEATURE_SHARED_BUSYBOX" = y; then
-    EXE="$sharedlib_dir/busybox_unstripped"
-    try $CC $CFLAGS $LDFLAGS \
-	    -o $EXE \
-	    $SORT_COMMON \
-	    $SORT_SECTION \
-	    $GC_SECTIONS \
-	    $START_GROUP $O_FILES $END_GROUP \
-	    -L"$sharedlib_dir" -lbusybox \
-	    $l_list \
-	    `INFO_OPTS` \
-    || {
-	echo "Linking $EXE failed"
-	cat $EXE.out
-	exit 1
-    }
-    $STRIP -s --remove-section=.note --remove-section=.comment $EXE -o "$sharedlib_dir/busybox"
-    echo "busybox linked against libbusybox: $sharedlib_dir/busybox"
-fi
-
-if test "$CONFIG_FEATURE_INDIVIDUAL" = y; then
-    echo "Linking individual applets against libbusybox (see $sharedlib_dir/*)"
-    gcc -DNAME_MAIN -E -include include/autoconf.h include/applets.h \
-    | grep -v "^#" \
-    | grep -v "^ *$" \
-    > applet_lst.tmp
-    while read name main junk; do
-
-	echo "\
-void lbb_prepare(const char *applet, char **argv);
-int $main(int argc, char **argv);
-
-int main(int argc, char **argv)
-{
-	lbb_prepare(\"$name\", argv);
-	return $main(argc, argv);
-}
-" >"$sharedlib_dir/applet.c"
-
-	EXE="$sharedlib_dir/$name"
-	try $CC $CFLAGS $LDFLAGS "$sharedlib_dir/applet.c" \
-		-o $EXE \
-		$SORT_COMMON \
-		$SORT_SECTION \
-		$GC_SECTIONS \
-		-L"$sharedlib_dir" -lbusybox \
-		-Wl,--warn-common \
-	|| {
-	    echo "Linking $EXE failed"
-	    cat $EXE.out
-	    exit 1
-	}
-	rm -- "$sharedlib_dir/applet.c" $EXE.out
-	$STRIP -s --remove-section=.note --remove-section=.comment $EXE
-	# Let user see that we do something - list the names of created binaries:
-	echo "$EXE"
-
-    done <applet_lst.tmp
-fi
-
-# libbusybox.so is needed only for -lbusybox at link time,
-# it is not needed at runtime. Deleting to reduce confusion.
-rm "$sharedlib_dir"/libbusybox.so 2>/dev/null
-exit 0 # or else we may confuse make
+os.system(" ".join([cc, '-o', out, cflags, ldflags, *obj, *lib]))
